### PR TITLE
[dynamic-shapes] basic slicing support with dynamic shape result

### PR DESCRIPTION
**Pull request by mattjj**
This PR also adds `__iter__` support for dynamically-shaped arrays.

### CODE REVIEW

File: jax/_src/lax/slicing.py [hunk 1]
```diff
@@ -684,7 +689,7 @@ def index_in_dim(operand: Array, index: int, axis: int = 0,
 def dynamic_slice_in_dim(operand: Array, start_index: Array,
                          slice_size: int, axis: int = 0) -> Array:
   """Convenience wrapper around dynamic_slice applying to one dimension."""
-  start_indices = [lax._zero(start_index)] * operand.ndim
+  start_indices = [0] * operand.ndim
```
Comment (sharadmv) [inline jax/_src/lax/slicing.py:692]: I understand why we are doing this now, but why didn't we do it before?

File: jax/_src/lax/slicing.py [hunk 2]
```diff
@@ -902,22 +909,49 @@ def _dynamic_slice_batching_rule(batched_args, batch_dims, *, slice_sizes):
...
     weak_type_rule=_argnum_weak_type(0))
 ad.primitive_jvps[dynamic_slice_p] = _dynamic_slice_jvp
 ad.primitive_transposes[dynamic_slice_p] = _dynamic_slice_transpose_rule
 batching.primitive_batchers[dynamic_slice_p] = _dynamic_slice_batching_rule
+pe.custom_staging_rules[dynamic_slice_p] = _dynamic_slice_staging_rule
+core.custom_typechecks[dynamic_slice_p] = _dynamic_slice_typecheck_rule
 
 def _dynamic_slice_lower(ctx, x, *start_indices, slice_sizes):
   aval_out, = ctx.avals_out
+  breakpoint()
```
Comment (sharadmv) [inline jax/_src/lax/slicing.py:949]: Delete please!

File: jax/_src/lax/slicing.py [hunk 3]
```diff
@@ -2058,6 +2092,9 @@ def _dynamic_slice_indices(operand, start_indices: Any):
     if isinstance(i, (int, np.integer)) and core.is_constant_dim(d):
       result.append(lax.convert_element_type(i + d, _dtype(i)) if i < 0 else i)
     else:
-      d = lax.convert_element_type(core.dimension_as_value(d), _dtype(i))
-      result.append(lax.select(i < 0, i + d, i))
+      d = lambda: lax.convert_element_type(core.dimension_as_value(d), _dtype(i))
```
Comment (sharadmv) [inline jax/_src/lax/slicing.py:2095]: Could add comment here saying you're trying to avoid unnecessary staging

**Review by sharadmv — APPROVED** [review]
Woo nice!

File: jax/_src/lax/slicing.py (hunk 1 above)
Comment (mattjj) [reply jax/_src/lax/slicing.py:692]: Indeed... 🤷‍♂️ 

Hope past versions of ourselves weren't being too clever.

File: jax/_src/lax/slicing.py (hunk 2 above)
Comment (mattjj) [reply jax/_src/lax/slicing.py]: What.

File: jax/_src/lax/slicing.py (hunk 3 above)
Comment (mattjj) [reply jax/_src/lax/slicing.py]: Done.

File: jax/_src/lax/slicing.py (hunk 1 above)
Comment (mattjj) [reply jax/_src/lax/slicing.py:692]: Revising...

**Pull request merged**